File: internal/looptui/layouts.go
package looptui

import "fmt"

type paneLayout struct {
	Rows int
	Cols int
}

var paneLayouts = []paneLayout{
	{Rows: 1, Cols: 1},
	{Rows: 1, Cols: 2},
	{Rows: 1, Cols: 3},
	{Rows: 1, Cols: 4},
	{Rows: 2, Cols: 2},
	{Rows: 2, Cols: 3},
	{Rows: 2, Cols: 4},
	{Rows: 3, Cols: 3},
	{Rows: 3, Cols: 4},
	{Rows: 4, Cols: 4},
}

func (l paneLayout) Capacity() int {
	if l.Rows < 1 || l.Cols < 1 {
		return 1
	}
	return l.Rows * l.Cols
}

func (l paneLayout) Label() string {
	return fmt.Sprintf("%dx%d", l.Rows, l.Cols)
}

func normalizeLayoutIndex(idx int) int {
	if len(paneLayouts) == 0 {
		return 0
	}
	for idx < 0 {
		idx += len(paneLayouts)
	}
	return idx % len(paneLayouts)
}

func layoutIndexFor(rows, cols int) int {
	for i, layout := range paneLayouts {
		if layout.Rows == rows && layout.Cols == cols {
			return i
		}
	}
	return 0
}

func layoutCellSize(layout paneLayout, width, height, gap int) (int, int) {
	if layout.Rows < 1 || layout.Cols < 1 {
		layout = paneLayout{Rows: 1, Cols: 1}
	}
	if gap < 0 {
		gap = 0
	}
	cellWidth := (width - ((layout.Cols - 1) * gap)) / layout.Cols
	cellHeight := (height - ((layout.Rows - 1) * gap)) / layout.Rows
	return cellWidth, cellHeight
}

func layoutFits(layout paneLayout, width, height, gap, minCellWidth, minCellHeight int) bool {
	cellWidth, cellHeight := layoutCellSize(layout, width, height, gap)
	return cellWidth >= minCellWidth && cellHeight >= minCellHeight
}

func fitPaneLayout(requested paneLayout, width, height, gap, minCellWidth, minCellHeight int) paneLayout {
	if requested.Rows < 1 || requested.Cols < 1 {
		requested = paneLayout{Rows: 1, Cols: 1}
	}
	if gap < 0 {
		gap = 0
	}
	if minCellWidth < 1 {
		minCellWidth = 1
	}
	if minCellHeight < 1 {
		minCellHeight = 1
	}

	best := paneLayout{Rows: 1, Cols: 1}
	bestScore := -1

	for _, candidate := range paneLayouts {
		if candidate.Rows > requested.Rows || candidate.Cols > requested.Cols {
			continue
		}
		if !layoutFits(candidate, width, height, gap, minCellWidth, minCellHeight) {
			continue
		}
		score := candidate.Capacity()*100 - (requested.Rows-candidate.Rows)*4 - (requested.Cols-candidate.Cols)*3
		if score > bestScore {
			best = candidate
			bestScore = score
		}
	}
	if bestScore >= 0 {
		return best
	}

	for i := len(paneLayouts) - 1; i >= 0; i-- {
		candidate := paneLayouts[i]
		if layoutFits(candidate, width, height, gap, minCellWidth, minCellHeight) {
			return candidate
		}
	}

	return best
}

File: rust/crates/forge-tui/src/layouts.rs
//! Loop TUI pane layout helpers.
//!
//! Parity port of `internal/looptui/layouts.go`.

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct PaneLayout {
    pub rows: i32,
    pub cols: i32,
}

pub const PANE_LAYOUTS: [PaneLayout; 10] = [
    PaneLayout { rows: 1, cols: 1 },
    PaneLayout { rows: 1, cols: 2 },
    PaneLayout { rows: 1, cols: 3 },
    PaneLayout { rows: 1, cols: 4 },
    PaneLayout { rows: 2, cols: 2 },
    PaneLayout { rows: 2, cols: 3 },
    PaneLayout { rows: 2, cols: 4 },
    PaneLayout { rows: 3, cols: 3 },
    PaneLayout { rows: 3, cols: 4 },
    PaneLayout { rows: 4, cols: 4 },
];

impl PaneLayout {
    #[must_use]
    pub fn capacity(self) -> i32 {
        if self.rows < 1 || self.cols < 1 {
            return 1;
        }
        self.rows.saturating_mul(self.cols).max(1)
    }

    #[must_use]
    pub fn label(self) -> String {
        format!("{}x{}", self.rows, self.cols)
    }
}

#[must_use]
pub fn normalize_layout_index(mut idx: i32) -> usize {
    if PANE_LAYOUTS.is_empty() {
        return 0;
    }
    while idx < 0 {
        idx += PANE_LAYOUTS.len() as i32;
    }
    (idx as usize) % PANE_LAYOUTS.len()
}

#[must_use]
pub fn layout_index_for(rows: i32, cols: i32) -> usize {
    for (i, layout) in PANE_LAYOUTS.iter().enumerate() {
        if layout.rows == rows && layout.cols == cols {
            return i;
        }
    }
    0
}

#[must_use]
pub fn layout_cell_size(
    mut layout: PaneLayout,
    width: i32,
    height: i32,
    mut gap: i32,
) -> (i32, i32) {
    if layout.rows < 1 || layout.cols < 1 {
        layout = PaneLayout { rows: 1, cols: 1 };
    }
    if gap < 0 {
        gap = 0;
    }

    let cell_width = (width - ((layout.cols - 1) * gap)) / layout.cols;
    let cell_height = (height - ((layout.rows - 1) * gap)) / layout.rows;
    (cell_width, cell_height)
}

#[must_use]
pub fn layout_fits(
    layout: PaneLayout,
    width: i32,
    height: i32,
    gap: i32,
    min_cell_width: i32,
    min_cell_height: i32,
) -> bool {
    let (cell_width, cell_height) = layout_cell_size(layout, width, height, gap);
    cell_width >= min_cell_width && cell_height >= min_cell_height
}

#[must_use]
pub fn fit_pane_layout(
    mut requested: PaneLayout,
    width: i32,
    height: i32,
    mut gap: i32,
    mut min_cell_width: i32,
    mut min_cell_height: i32,
) -> PaneLayout {
    if requested.rows < 1 || requested.cols < 1 {
        requested = PaneLayout { rows: 1, cols: 1 };
    }
    if gap < 0 {
        gap = 0;
    }
    if min_cell_width < 1 {
        min_cell_width = 1;
    }
    if min_cell_height < 1 {
        min_cell_height = 1;
    }

    let mut best = PaneLayout { rows: 1, cols: 1 };
    let mut best_score = -1i32;

    for candidate in PANE_LAYOUTS {
        if candidate.rows > requested.rows || candidate.cols > requested.cols {
            continue;
        }
        if !layout_fits(
            candidate,
            width,
            height,
            gap,
            min_cell_width,
            min_cell_height,
        ) {
            continue;
        }
        let score = candidate.capacity() * 100
            - (requested.rows - candidate.rows) * 4
            - (requested.cols - candidate.cols) * 3;
        if score > best_score {
            best = candidate;
            best_score = score;
        }
    }
    if best_score >= 0 {
        return best;
    }

    for candidate in PANE_LAYOUTS.iter().rev() {
        if layout_fits(
            *candidate,
            width,
            height,
            gap,
            min_cell_width,
            min_cell_height,
        ) {
            return *candidate;
        }
    }

    best
}

#[cfg(test)]
mod tests {
    use super::{fit_pane_layout, layout_cell_size, PaneLayout};

    #[test]
    fn fit_pane_layout_degrades_when_space_too_small() {
        let requested = PaneLayout { rows: 4, cols: 4 };
        let actual = fit_pane_layout(requested, 120, 30, 2, 44, 10);
        assert!(actual.rows < requested.rows || actual.cols < requested.cols);
        assert!(actual.capacity() >= 1);
    }

    #[test]
    fn fit_pane_layout_keeps_requested_when_large_enough() {
        let requested = PaneLayout { rows: 2, cols: 2 };
        let actual = fit_pane_layout(requested, 240, 80, 2, 44, 10);
        assert_eq!(actual, requested);
    }

    #[test]
    fn fit_pane_layout_preserves_columns_when_height_is_limited() {
        let requested = PaneLayout { rows: 4, cols: 4 };
        let actual = fit_pane_layout(requested, 220, 26, 1, 44, 10);
        let expected = PaneLayout { rows: 2, cols: 4 };
        assert_eq!(actual, expected);
    }

    #[test]
    fn layout_cell_size_matches_go_math() {
        let (cell_width, cell_height) =
            layout_cell_size(PaneLayout { rows: 2, cols: 4 }, 220, 26, 1);
        assert_eq!(cell_width, 54);
        assert_eq!(cell_height, 12);
    }
}
